File: src/didery/static/transcrypt/components/tabledtab.py
# ================================================== #
#                     TABLED TAB                     #
# ================================================== #
# Author: Brady Hammond                              #
# Created: 05/08/2018                                #
# Last Edited:                                       #
# Last Edited By:                                    #
# ================================================== #
#                      IMPORTS                       #
# ================================================== #

import didery.static.transcrypt.components.tab as tab

# ================================================== #
#                  CLASS DEFINITIONS                 #
# ================================================== #

class TabledTab(tab.Tab):
    """
    Base class for tabs in the dashboard interface. Uses table and details view.
    """
    def __init__(self):
        """
        Initialize TabledTab object.
        """
        super().__init__()
        self.table = None
        self.setup_table()
        self.copiedDetails = ""
        self._detailsId = self.DataTab + "DetailsCodeBlock"
        self._copiedId = self.DataTab + "CopiedCodeBlock"
        self._copyButtonId = self.DataTab + "CopyButton"
        self._clearButtonId = self.DataTab + "ClearButton"

    # ============================================== #

    def setup_table(self):
        """
        Sets up table object.
        """
        self.table = Table([])

    # ============================================== #

    def _copyDetails(self):
        self.copiedDetails = self.table.detailSelected

    # ============================================== #

    def _getRows(self):
        return jQuery("[data-tab='{0}'].tab table > tbody > tr".format(self.DataTab))

    # ============================================== #

    def _getLabel(self):
        return jQuery(".menu a[data-tab='{0}'] .ui.label".format(self.DataTab))

    # ============================================== #

    def _clearCopy(self):
        self.copiedDetails = ""

    # ============================================== #

    def menu_item(self):
        return m(self._menu, self._menu_attrs,
                 m("div.menu-item-text", self.Name),
                 m(self.Icon),
                 m("div.ui.label.small.menu-item-number", "{0}/{1}".format(self.table.shown, self.table.total))
                 )

    # ============================================== #

    def main_view(self):
        return m("div",
                 m("div.table-container", m(self.table.view)),
                 m("div.ui.hidden.divider"),
                 m("div.ui.two.cards", {"style": "height: 45%;"},
                   m("div.ui.card",
                     m("div.content.small-header",
                       m("div.header",
                         m("span", "Details"),
                         m("span.ui.mini.right.floated.button", {"onclick": self._copyDetails, "id": self._copyButtonId},
                           "Copy")
                         )
                       ),
                     m("pre.content.code-block", {"id": self._detailsId},
                       self.table.detailSelected
                       )
                     ),
                   m("div.ui.card",
                     m("div.content.small-header",
                       m("div.header",
                         m("span", "Copied"),
                         m("span.ui.mini.right.floated.button", {"onclick": self._clearCopy, "id": self._clearButtonId},
                           "Clear")
                         )
                       ),
                     m("pre.content.code-block", {"id": self._copiedId},
                       self.copiedDetails
                       )
                     )
                   )
                 )

# ================================================== #
#                        EOF                         #
# ================================================== #

File: src/didery/static/js/components/tabledtab.js
let tab = require("./tab");

module.exports = {
    TabledTab: class TTab extends tab.Tab {
        constructor() {
            super();

            this.table = null;
            this.setup_table();
            this.copiedDetails = "";
            this._detailsId = this.DataTab + "DetailsCodeBlock";
            this._copiedId = this.DataTab + "CopiedCodeBlock";
            this._copyButtonId = this.DataTab + "CopyButton";
            this._clearButtonId = this.DataTab + "ClearButton";
        }

        setup_table() {
            /*
             * Sets up table object.
             */
            this.table = new Table([]);
        }

        _copyDetails() {
            this.copiedDetails = this.table.detailSelected;
        }

        _getRows() {
            let selector = "[data-tab='" + this.DataTab + "'].tab table > tbody > tr";
            return $(selector);
        }

        _getLabel() {
            let selector = ".menu a[data-tab='" + this.DataTab + "'] .ui.label";
            return $(selector);
        }

        _clearCopy() {
            this.copiedDetails = "";
        }

        menu_item() {
            return m(this._menu, this._menu_attrs,
                m("div.menu-item-text", this.Name),
                m(this.Icon),
                m("div.ui.label.small.menu-item-number", "{0}/{1}".format(this.table.shown, this.table.total))
            );
        }

        main_view() {
            return m("div",
                m("div.table-container", m(this.table.view)),
                m("div.ui.hidden.divider"),
                m("div.ui.two.cards", {"style": "height: 45%;"},
                    m("div.ui.card",
                        m("div.content.small-header",
                            m("div.header",
                                m("span", "Details"),
                                m("span.ui.mini.right.floated.button", {
                                        "onclick": this._copyDetails,
                                        "id": this._copyButtonId
                                    },
                                    "Copy")
                            )
                        ),
                        m("pre.content.code-block", {"id": this._detailsId},
                            this.table.detailSelected
                        )
                    ),
                    m("div.ui.card",
                        m("div.content.small-header",
                            m("div.header",
                                m("span", "Copied"),
                                m("span.ui.mini.right.floated.button", {
                                        "onclick": this._clearCopy,
                                        "id": this._clearButtonId
                                    },
                                    "Clear")
                            )
                        ),
                        m("pre.content.code-block", {"id": this._copiedId},
                            this.copiedDetails
                        )
                    )
                )
            )
        }
    }
};